hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/SimpleRegionNormalizer.java

```diff
@@ -18,126 +17,436 @@
+    }
+
+    LOG.debug("Table {}, total aggregated regions size: {} and average region size {}", table,
+      totalSizeMb, avgRegionSize);
+    return avgRegionSize;
+  }
+
+  /**
+   * Determine if a {@link RegionInfo} should be considered for a merge operation.
+   */
+  private boolean skipForMerge(final RegionStates regionStates, final RegionInfo regionInfo) {
+    final RegionState state = regionStates.getRegionState(regionInfo);
+    final String name = regionInfo.getEncodedName();
+    return
+      logTraceReason(
+        () -> state == null,
+        "skipping merge of region {} because no state information is available.", name)
+        || logTraceReason(
+          () -> !Objects.equals(state.getState(), RegionState.State.OPEN),
+          "skipping merge of region {} because it is not open.", name)
+        || logTraceReason(
+          () -> !isOldEnoughForMerge(regionInfo),
+          "skipping merge of region {} because it is not old enough.", name)
+        || logTraceReason(
+          () -> !isLargeEnoughForMerge(regionInfo),
+          "skipping merge region {} because it is not large enough.", name);
+  }
+
+  /**
+   * Computes the merge plans that should be executed for this table to converge average region
+   * towards target average or target region count
+   * @param table table to normalize
+   * @return list of merge normalization plans
+   */
+  private List<NormalizationPlan> computeMergeNormalizationPlans(TableName table) {
+    final RegionStates regionStates = masterServices.getAssignmentManager().getRegionStates();
+    final List<RegionInfo> tableRegions = regionStates.getRegionsOfTable(table);
+
+    if (tableRegions.size() < minRegionCount) {
+      LOG.debug("Table {} has {} regions, required min number of regions for normalizer to run"
+        + " is {}, not computing merge plans.", table, tableRegions.size(), minRegionCount);
+      return Collections.emptyList();
+    }
+
+    final double avgRegionSize = getAverageRegionSize(tableRegions);
+    LOG.debug("Table {}, average region size: {}. Computing normalization plan for table: {}, "
+        + "number of regions: {}.",
+      table, avgRegionSize, table, tableRegions.size());
+
+    // The list of regionInfo from getRegionsOfTable() is ordered by regionName.
+    // regionName does not necessary guarantee the order by STARTKEY (let's say 'aa1', 'aa1!',
+    // in order by regionName, it will be 'aa1!' followed by 'aa1').
+    // This could result in normalizer merging non-adjacent regions into one and creates overlaps.
+    // In order to avoid that, sort the list by RegionInfo.COMPARATOR.
+    tableRegions.sort(RegionInfo.COMPARATOR);
+    final List<NormalizationPlan> plans = new ArrayList<>();
+    for (int candidateIdx = 0; candidateIdx < tableRegions.size() - 1; candidateIdx++) {
+      final RegionInfo hri = tableRegions.get(candidateIdx);
+      final RegionInfo hri2 = tableRegions.get(candidateIdx + 1);
```

Comment:
nit: Better var names will be `hri` and `hriNext` or `hriAdjacent`? Same goes for `regionSize` and `regionSize2`.

Reply:
Indeed. Let me clean this up too.